optuna/integration/wandb.py

```diff
@@ -124,12 +134,52 @@ def __call__(self, study: optuna.study.Study, trial: optuna.trial.FrozenTrial) -
                 names = [*self._metric_name]
 
         metrics = {name: value for name, value in zip(names, trial.values)}
+
+        if self._as_multirun:
+            metrics["trial_number"] = trial.number
+
         attributes = {"direction": [d.name for d in study.directions]}
 
-        wandb.config.update(attributes)
-        wandb.log({**trial.params, **metrics}, step=trial.number)
+        step = trial.number if wandb.run else None
+        run = wandb.run
 
-    def _initialize_run(self) -> None:
-        """Initializes Weights & Biases run."""
+        # When an user doesn't use `track_in_wandb` decorator.
+        # Might create an extra run if an user logs in wandb but doesn't use the decorator.
+
+        if not run:
+            run = self._initialize_run()
+            run.name = f"trial/{trial.number}/{run.name}"
 
-        wandb.init(**self._wandb_kwargs)
+        run.log({**trial.params, **metrics}, step=step)
+
+        if self._as_multirun:
+            run.config.update({**attributes, **trial.params})
+            run.tags = tuple(self._wandb_kwargs.get("tags", ())) + (study.study_name,)
+            run.finish()
+        else:
+            run.config.update(attributes)
+
+    @experimental("3.0.0")
+    def track_in_wandb(self, func: ObjectiveFuncType) -> ObjectiveFuncType:
```

Comment:
TODO: we need to add documentation here like https://optuna.readthedocs.io/en/stable/reference/generated/optuna.integration.MLflowCallback.html#optuna.integration.MLflowCallback. But it is completely okay to add it as a follow-up.